dask/distributed/scheduler.py

```diff
@@ -717,5 +717,33 @@ def _get_workers(self, header, payload):
     def _heartbeat(self, header, payload):
         with logerrors():
             log(self.address_to_clients, "Heartbeat", header)
+            payload = pickle.loads(payload)
             address = header['address']
+
+            if address not in self.workers:
+                self.available_workers.put(address)
+
+            self.workers[address] = payload
             self.workers[address]['last-seen'] = datetime.utcnow()
+
+    def prune_workers(self, timeout=20):
+        """
+        Remove workers from scheduler that have not sent a heartbeat in
+        `timeout` seconds.
+        """
+        now = datetime.utcnow()
+        remove = []
+        with self.lock:
```

Comment:
Really? In `Scheduler.__init__` I see  `self.lock = Lock()` where `Lock` is imported from `threading`. So wouldn't this lock all threads created by the scheduler? Including any threads are mutating `self.workers`?